cmd/snap-seccomp/main_test.go

```diff
@@ -0,0 +1,521 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main_test
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strconv"
+	"strings"
+	"testing"
+	"unsafe"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/mvo5/libseccomp-golang"
+
+	// forked from "golang.org/x/net/bpf"
+	// until https://github.com/golang/go/issues/20556
+	"github.com/mvo5/net/bpf"
+
+	main "github.com/snapcore/snapd/cmd/snap-seccomp"
+)
+
+// Hook up check.v1 into the "go test" runner
+func Test(t *testing.T) { TestingT(t) }
+
+type snapSeccompSuite struct{}
+
+var _ = Suite(&snapSeccompSuite{})
+
+func goArchToScmpArch(goarch string) uint32 {
+	switch goarch {
+	case "386":
+		return main.ScmpArchX86
+	case "amd64":
+		return main.ScmpArchX86_64
+	case "arm":
+		return main.ScmpArchARM
+	case "arm64":
+		return main.ScmpArchAARCH64
+	case "ppc64le":
+		return main.ScmpArchPPC64LE
+	case "s390x":
+		return main.ScmpArchS390X
+	case "ppc":
+		return main.ScmpArchPPC
+	}
+	panic(fmt.Sprintf("cannot map goarch %q to a seccomp arch", goarch))
+}
+
+func decodeBpfFromFile(p string) ([]bpf.Instruction, error) {
+	var ops []bpf.Instruction
+	var rawOp bpf.RawInstruction
+
+	r, err := os.Open(p)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+
+	for {
+		err = binary.Read(r, nativeEndian(), &rawOp)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		ops = append(ops, rawOp.Disassemble())
+	}
+
+	return ops, nil
+}
+
+func parseBpfInput(s string) (*main.SeccompData /* *seccompData */, error) {
+	// syscall;arch;arg1,arg2...
+	l := strings.Split(s, ";")
+
+	sc, err := seccomp.GetSyscallFromName(l[0])
+	if err != nil {
+		return nil, err
+	}
+
+	var arch uint32
+	if len(l) < 2 || l[1] == "native" {
+		arch = goArchToScmpArch(runtime.GOARCH)
+	} else {
+		arch = goArchToScmpArch(l[1])
+	}
+
+	var syscallArgs [6]uint64
+	if len(l) > 2 {
+		args := strings.Split(l[2], ",")
+		for i := range args {
+			// init with random number argument
+			syscallArgs[i] = (uint64)(rand.Uint32())
+			// override if the test specifies a specific number
+			if nr, err := strconv.ParseUint(args[i], 10, 64); err == nil {
+				syscallArgs[i] = nr
+			} else if nr, ok := main.SeccompResolver[args[i]]; ok {
+				syscallArgs[i] = nr
+			}
+		}
+	}
+	sd := &main.SeccompData{}
+	sd.SetArch(arch)
+	sd.SetNr(sc)
+	sd.SetArgs(syscallArgs)
+	return sd, nil
+}
+
+// Endianness detection.
+func nativeEndian() binary.ByteOrder {
+	// Credit matt kane, taken from his gosndfile project.
+	// https://groups.google.com/forum/#!msg/golang-nuts/3GEzwKfRRQw/D1bMbFP-ClAJ
+	// https://github.com/mkb218/gosndfile
+	var i int32 = 0x01020304
+	u := unsafe.Pointer(&i)
+	pb := (*byte)(u)
+	b := *pb
+	isLittleEndian := b == 0x04
+
+	if isLittleEndian {
+		return binary.LittleEndian
+	} else {
+		return binary.BigEndian
+	}
+}
+
+func simulateBpf(c *C, seccompWhitelist, bpfInput string, expected int) {
```

Comment:
Personally, I think 30 seconds is acceptable, but note my last comment-- I think we can make it faster.

What is neat about your approach is 'yes' it is very close and we're doing it all in unit tests rather than relying on this only in the integration spread tests. Yay! :)